File: source_code/Calculations/Max_KZ_Time.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.IO;
using System.Linq;
using rst_operations;

namespace DynStabSpace;

public class Max_KZ_Time
{
	private string tmp_root = Environment.GetFolderPath(Environment.SpecialFolder.Personal) + $"\\DynStabSpace\\{DateTime.Now:yyyy-MM-dd HH-mm-ss} Предельное время КЗ";

	private BackgroundWorker _w;

	private ObservableCollection<RgmsInfo> RgmsInfo;

	private ObservableCollection<ScnsInfo> ScnsInfo;

	private List<VrnInfo> VrnInf;

	private FileInfo Rems;

	private double CrtTimePrecision;

	private double CrtTimeMax;

	public string root => tmp_root;

	public int Max => RgmsInfo.Count() * VrnInf.Where((VrnInfo k) => !k.Deactive).Count() * ScnsInfo.Count() + 1;

	public Max_KZ_Time(object sender, ObservableCollection<RgmsInfo> rgms, ObservableCollection<ScnsInfo> scns, List<VrnInfo> vrns, FileInfo rems, double time_precision, double max_time)
	{
		if (rgms.Count() == 0 || scns.Count() == 0 || max_time == 0.0)
		{
			throw new InitialDataException("Не заданы все исходные данные для определения предельного времени отключения КЗ!" + Environment.NewLine + Environment.NewLine + "Результаты проверки:" + Environment.NewLine + $"Загружено расчетных режимов - {rgms.Count()}{Environment.NewLine}" + $"Загружено аварийных процесслов - {scns.Count()}{Environment.NewLine}" + $"Максимальное время отключения КЗ составляет - {max_time}");
		}
		RgmsInfo = rgms;
		ScnsInfo = scns;
		VrnInf = vrns;
		Rems = rems;
		CrtTimePrecision = time_precision;
		CrtTimeMax = max_time;
		_w = sender as BackgroundWorker;
		Directory.CreateDirectory(tmp_root);
	}

	public List<CrtTimeResults> Calc()
	{
		int num = 0;
		List<CrtTimeResults> list = new List<CrtTimeResults>();
		foreach (RgmsInfo item in RgmsInfo)
		{
			List<CrtShems> list2 = new List<CrtShems>();
			foreach (VrnInfo item2 in VrnInf.Where((VrnInfo k) => !k.Deactive))
			{
				List<CrtTimes> list3 = new List<CrtTimes>();
				CrtShems crtShems = new CrtShems();
				crtShems.ShemeName = item2.Name;
				foreach (ScnsInfo item3 in ScnsInfo)
				{
					RastrOperations rastrOperations = new RastrOperations();
					rastrOperations.Load(item.Name);
					rastrOperations.Load(item3.Name);
					crtShems.IsStable = ((item2.Id == -1) ? rastrOperations.rgm() : rastrOperations.ApplyVariant(item2.Num, Rems.Name));
					if (crtShems.IsStable)
					{
						list3.Add(new CrtTimes
						{
							ScnName = item3.name,
							CrtTime = rastrOperations.FindCrtTime(CrtTimePrecision, CrtTimeMax)
						});
						num++;
						_w.ReportProgress(num);
					}
				}
				crtShems.Times = list3;
				list2.Add(crtShems);
			}
			list.Add(new CrtTimeResults
			{
				RgName = item.name,
				CrtShems = list2
			});
		}
		return list;
	}
}

File: python_dss/calculations/max_kz_time.py
"""
Расчет предельного времени отключения КЗ
"""

import os
from datetime import datetime
from pathlib import Path
from typing import List, Optional, Callable
from models import (
    RgmsInfo, ScnsInfo, VrnInfo, CrtTimeResults, CrtShems, CrtTimes
)
from rastr_operations import RastrOperations
from utils.exceptions import InitialDataException


class MaxKZTimeCalc:
    """Расчет предельного времени отключения КЗ"""
    
    def __init__(self, progress_callback: Optional[Callable[[int], None]],
                 rgms: List[RgmsInfo], scns: List[ScnsInfo], vrns: List[VrnInfo],
                 rems_path: Optional[str], time_precision: float, max_time: float):
        """
        Инициализация расчета предельного времени КЗ
        
        Args:
            progress_callback: Функция обратного вызова для обновления прогресса
            rgms: Список расчетных режимов
            scns: Список аварийных процессов
            vrns: Список вариантов (ремонтных схем)
            rems_path: Путь к файлу ремонтных схем
            time_precision: Точность расчета (секунды)
            max_time: Максимальное время отключения КЗ (секунды)
        """
        if not rgms or not scns or max_time == 0.0:
            error_msg = "Не заданы все исходные данные для определения предельного времени отключения КЗ!\n\n"
            error_msg += f"Результаты проверки:\n"
            error_msg += f"Загружено расчетных режимов - {len(rgms)}\n"
            error_msg += f"Загружено аварийных процесслов - {len(scns)}\n"
            error_msg += f"Максимальное время отключения КЗ составляет - {max_time}\n"
            raise InitialDataException(error_msg)
        
        self._progress_callback = progress_callback
        self._rgms = rgms
        self._scns = scns
        self._vrns = [v for v in vrns if not v.deactive]
        self._rems_path = rems_path
        self._time_precision = time_precision
        self._max_time = max_time
        
        # Создание папки для результатов
        self._root = Path.home() / "DynStabSpace" / f"{datetime.now():%Y-%m-%d %H-%M-%S} Предельное время КЗ"
        self._root.mkdir(parents=True, exist_ok=True)
    
    @property
    def root(self) -> str:
        """Путь к папке с результатами"""
        return str(self._root)
    
    @property
    def max(self) -> int:
        """Максимальное количество шагов расчета"""
        return len(self._rgms) * len(self._vrns) * len(self._scns) + 1
    
    def calc(self) -> List[CrtTimeResults]:
        """Выполнение расчета"""
        progress = 0
        results = []
        
        for rgm in self._rgms:
            crt_shems_list = []
            
            for vrn in self._vrns:
                times_list = []
                crt_shem = CrtShems(sheme_name=vrn.name, is_stable=False, times=[])
                
                for scn in self._scns:
                    rastr = RastrOperations()
                    rastr.load(rgm.name)
                    rastr.load(scn.name)
                    
                    # Применение варианта
                    if vrn.id == -1:
                        is_stable = rastr.rgm()
                    else:
                        is_stable = rastr.apply_variant(vrn.num, self._rems_path)
                    
                    crt_shem.is_stable = is_stable
                    
                    if is_stable:
                        # Поиск критического времени
                        crt_time = rastr.find_crt_time(self._time_precision, self._max_time)
                        
                        times_list.append(CrtTimes(
                            scn_name=Path(scn.name).stem,
                            crt_time=crt_time
                        ))
                        
                        progress += 1
                        if self._progress_callback:
                            self._progress_callback(progress)
                
                crt_shem.times = times_list
                crt_shems_list.append(crt_shem)
            
            results.append(CrtTimeResults(
                rg_name=Path(rgm.name).stem,
                crt_shems=crt_shems_list
            ))
        
        return results

